### PR TITLE
3D Explosion Blast Forces and 3D Splash Damage

### PR DESCRIPTION
**Pull request by spartahawk**
Lots of scenarios to test here. Maybe some people on the forum can help us out by trying this in another alpha release.

3D explosion blast forces and 3D splash damage implemented. Somewhat hacky in the vertical plane, but it works. And adjustments will likely need to be made to the explosion blast power to adjust for the 3D-distributed damage.

**If you shoot a meteor at the floor below a hovering ghost (or any flying creature), it will be pushed upward by the blast. I haven't tried grenades yet.**

I used the built-in `sqrt()` function from math.c, and square roots are expensive, so there's a chance this could cause lag in large fights with explosions. But it only has to do it once for each creature, so that's nothing really.

~~Also, the bug where if the explosion happened "within" a lump in the floor/wall (due to dungeon "wibble") and got "blocked" by the solid part of the lump it was surrounded by is hopefully adequately squashed.~~(irrelevant to this pull request now)

### CODE REVIEW

**Comment by spartahawk** [comment]
Once it's confirmed that the floor/wall lumps are no longer blocking explosions, I'll remove those custom logs from the code in the line_of_sight_3d function.

**Comment by spartahawk** [comment]
We should look at how Word Of Power behaves as well. It's affected by this too, right? @Loobinex this uses the code that you so-far confirmed fixed the floor explosion bug, so if you're satisfied with that remember to change the values in rules.cfg back to test the 3D explosions.

Let me know if the vertical forces are proportionately too weak compared to the horizontal forces.

**Comment by Loobinex** [comment]
I'll test later today, but you do realize that the values changed in rules.cfg was to make friendly fire equally effective as enemy fire, so as to easily test.
If the current values are too much, then they are also too much for regular combat after the friendly fire is reduced again.

**Comment by spartahawk** [comment]
I had a game crash. I don't know if it was unrelated or not. I don't know if I have seen these before. But I haven't had a lot of experience looking at logs:

_Error: get_creature_model_graphics: Invalid model 32 graphics sequence 16
Error: get_creature_sound: Trying to get sound for undefined creature type 32_

And I'm getting both of these logs, but I think they are legitimate rather than wrong. It was during a big fight going on in two separate rooms, so I think the walls were correctly blocking the blast.

_Player cannot see through (137,59) due to linear path solid flags (downcount 4)
Player cannot see through (138,59) due to 3D line of sight (downcount 1)_

**Comment by Loobinex** [comment]
Crashes in this game unrelated to the the pathfinding bug are very rare. There's one more that sometimes happens when you watch an exploding dungeon heart, so there is a good chance that if a crash happened, it's because of a recent change.

However, one theory I have is that you used the cheat menu to add creatures to the game to fight, correct? That menu occasionally adds in non-existing units, and that can also crash the game AND give those error messages. Have you?

**Comment by spartahawk** [comment]
Yeah almost all the creatures were from the cheat menu.

**Comment by Loobinex** [comment]
FYI: Branch closed to separate out different features into different pull requests, and in testing I also had my game hang and later crash unrelated to the cheat menu.

File: src/creature_senses.h [hunk 1]
```diff
@@ -39,6 +39,7 @@ TbBool sibling_line_of_sight_ignoring_door(const struct Coord3d *prevpos,
 #define sibling_line_of_sight(prevpos, nextpos) sibling_line_of_sight_ignoring_door(prevpos, nextpos, INVALID_THING)
 
 TbBool line_of_sight_3d(const struct Coord3d *frpos, const struct Coord3d *topos);
+TbBool line_of_sight_3d_explosion(const struct Coord3d *frpos, const struct Coord3d *topos);
```
Comment (Loobinex) [inline src/creature_senses.h]: This line doesn't belong in this pull request.

**Comment by spartahawk** [comment]
You're referring to *TbBool sibling_line_of_sight_ignoring_door( )* right?


Are you saying it should have been deleted? I did not add it. Your comment
is deleted though so maybe you saw that.

**Comment by spartahawk** [comment]
Reopened pull request, since now that the commit to master from #25 was the only additional feature included in this request.

Crash by divide by zero appears to be resolved. Please confirm. I was most able to reproduce it by dropping several warlocks into a fight. It was seemingly caused when the vertical difference in the point of the explosion and the creature it could effect was zero.

**Comment by Schizz** [comment]
so, what's going on here? Is there still a problem with the line of sight?

**Comment by Loobinex** [comment]
sparta build a new feature, explosions that blast units away in 3 dimensions. So flying creatures get blown against the ceiling when you shoot the floor below them.

It's not merged yet, because I found an issue where if you shoot a meteor above a creature standing in water against a wall in this release, they will sink through the floor.

**Comment by Schizz** [comment]
well, is all of this really worth it? Since this wouldn't really change the game in any recognizable way for the player.

**Comment by Loobinex** [comment]
There's not much of a benefit to the player no. This one just caught spartas fancy.

**Comment by Schizz** [comment]
Maybe the game is not even meant to handle vertical forces. So there might appear more problems like the one you described, mostly with collision detection I guess.

**Comment by spartahawk** [comment]
There's a quick fix I'll try to apply in a little bit that I'm confident will resolve it, and after that if it's not working I'll move on and we can keep the code in whatever acceptable state we can include (like 2d only). I'm not so sure the issue is even that big of a deal. It's going to rarely happen anyway and it's a temporary thing. Creatures just get blown into the void for a second. Maybe that's a nice break after getting hit by a meteor.

But yeah, this was a distraction from more "important" stuff. Right now I'm looking at pathfinding game hangs.

**Comment by Loobinex** [comment]
The 3d line of sight was already put back last PR. So if we decide not to merge this PR, there's nothing else we need to do to get to an 'acceptable state'.

**Comment by spartahawk** [comment]
Let me know what you think of the behavior now. The following is pasted from my commit message:

Explosion blast pushes at angle to creature head rather than feet (see my argument below for the head rather than the creature's midsection). This fixes the issue of the blasts in liquid when a creature is up against a wall and they would briefly disappear.

A couple of results:

In battles, if any creature is within an explosion aimed at the torso of another creature, and any of those creatures' head is above the torso of the other creature, they will be pushed upward at that angle.

Also, it might just be my perspective but it seems like the creatures get pushed further horizontally. I don't know if this is true, but if it is, it might be due to lack of friction with the floor, though I doubt that is implemented in the game at all. If the movement is too far, we can tweak the explosion power in the config file, or adjust the proportions calculated in the code.

Creatures can be boosted or self-boosted (if they have a spell with an explosion) over guard posts that are next to water/lava.

I think that it looks/works best, especially as a possessed creature, for the blast to move the creature at the angle made from the explosion origin and your face (eyes)/head. If the blast pushes only as high as your midsection, you can see the blast but in some cases it won't push you at all because the line of sight to your midsection is blocked by something (like the edge of a guard post next to water, and you shoot down into the water, but you've stepped back on the guard post a bit).

**Comment by spartahawk** [comment]
Should bile demons not get pushed around by explosions?

**Comment by spartahawk** [comment]
Also, warlocks seem to kill themselves off by the explosions pretty quickly. Perhaps they should at least be immune to the damage of their own spell (but still get pushed by it perhaps)?

**Comment by Loobinex** [comment]
I have not tested yet, but some remarks:
* Warlocks might die easily now, because you're testing with 3 times the damage and radius.
* Bile demons are heavy, that's why they do not get pushed around
* It is defined in the config exactly how far each explosion should push creatures. We should stick to that value. Then if we want to push more or less, we can tweak the values in the config, but a mismatch between the config and actual range is not acceptable.

**Comment by spartahawk** [comment]
Thanks.

I will work on preventing blasts from pushing creatures over guard posts (and temple) next to water/lava.

Yeah this change doesn't override how far creatures get proportionately pushed according the the configs.

Bile demons always got pushed horizontally by word of power right? I've updated it so that creatures that are unaffected by wind don't get pushed vertically.

I've also changed the push angle to be based on the angle from the explosion to the midsection of the creature, although the line of sight remains from the explosion to the head (as argued for above). I think this creates are more realistic effect in the angle of movement while making sure creatures are still affected if any part of them is in the line of sight.

File: src/thing_effects.c [hunk 2]
```diff
@@ -1351,17 +1360,39 @@ TbBool explosion_affecting_thing(struct Thing *tngsrc, struct Thing *tngdst, con
...
+                                        if ((dist < param->num1) && crstat->affected_by_wind)
+                    */
+
                     if (dz != 0)
                     {
-                        tngdst->veloc_push_add.z.val += (dz / abs(dz)) * move_dist;
+                        // Only creatures, and those affected by wind (bile demons), get blasted upward
+                        if ( (tngdst->class_id == TCls_Creature && creature_stats_get_from_thing(tngdst)->affected_by_wind) )
+                        {
+                            tngdst->veloc_push_add.z.val += (dz / abs(dz)) * move_dist;
```
Comment (Loobinex) [inline src/thing_effects.c]: Just to be sure, this value can't be 0?

**Comment by Loobinex** [comment]
I think I like the vertical explosion at 0.2 or 0.3.
I see the units still turn invisible when shot at in water, and now I also noticed that flying creatures shot down will in fact get stuck behind guardposts and stuff and will not get up on their own.

Is it an idea to leave the bugs as they are for now, but move the verticalFactor from thing_effects.c to the [magic] section of the rules.cfg(as a percentage), and put both the default value and the value in the config to 0. That solves these issues, and gives people the option to play around with this all they like, and people can make campaigns(and hopefully in the future single maps_ with custom configs where it is possible to boost units over guard posts and whatnot?

**Comment by spartahawk** [comment]
That's a great way to go for now.

**Comment by spartahawk** [comment]
Good to merge. Double check my version numbers and squash & merge when you are satisfied.
